[OpenXR] ASSERTION FAILED: WebCore::GLContext::current() in ~OpenXRLayer

fujii opened the pull request:
#### 166ab722ba11ee040d21f6d70177bde3809c7f4b
<pre>
[OpenXR] ASSERTION FAILED: WebCore::GLContext::current() in ~OpenXRLayer
<a href="https://bugs.webkit.org/show_bug.cgi?id=304384">https://bugs.webkit.org/show_bug.cgi?id=304384</a>

Reviewed by Sergio Villar Senin.

An assertion failed in ~OpenXRLayer since there was no current GLContext. This
happened if the session was closed without any frames were requested.
makeContextCurrent() was called by `OpenXRCoordinator::beginFrame`. Thus, no
frame was requested, no current GLContext.

Call makeContextCurrent() in tryInitializeGraphicsBinding.

The test case of another PR will cover this case.
&lt;<a href="https://github.com/WebKit/WebKit/pull/55444">https://github.com/WebKit/WebKit/pull/55444</a>&gt;

* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::tryInitializeGraphicsBinding):
(WebKit::OpenXRCoordinator::beginFrame):
(WebKit::OpenXRCoordinator::endFrame):

Canonical link: <a href="https://commits.webkit.org/304668@main">https://commits.webkit.org/304668@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fa64fed29e52c90927b89a0bcdb3402099bca699

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136236 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8592 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47516 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143946 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9268 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8437 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104169 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139182 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6746 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122079 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85000 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6403 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4060 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4540 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115691 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146692 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8276 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40847 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112510 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8292 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6953 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112851 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6327 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118384 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20992 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8324 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36441 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8042 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71883 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8265 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8116 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

svillar reviewed (changes requested):
We need the make context current calls in both start and end frame because we cannot be sure that the UI process is using another one. 

I'd suggest to add special code to the destructor instead. We can early return if there is no current context. We'd likely have to add asserts in that case to verify that the fbos and the other resources we free have not been allocated.

fujii committed:
[OpenXR] ASSERTION FAILED: WebCore::GLContext::current() in ~OpenXRLayer
https://bugs.webkit.org/show_bug.cgi?id=304384

Reviewed by Sergio Villar Senin.

An assertion failed in ~OpenXRLayer since there was no current GLContext. This
happened if the session was closed without any frames were requested.
makeContextCurrent() was called by `OpenXRCoordinator::beginFrame`. Thus, no
frame was requested, no current GLContext.

Call makeContextCurrent() in tryInitializeGraphicsBinding.

The test case of another PR will cover this case.
<https://github.com/WebKit/WebKit/pull/55444>

* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::tryInitializeGraphicsBinding):
(WebKit::OpenXRCoordinator::beginFrame):
(WebKit::OpenXRCoordinator::endFrame):

Canonical link: https://commits.webkit.org/304668@main
